### PR TITLE
Add JSON schema validation for payloads

### PR DESCRIPTION
**Pull request by alexsavu**
Given the fairly complicated JSON payload we expect form clients due to various degrees of flexibility we offer a schema validation was put in place to offer certain restrictions on value types as well as nested objects

* tests for schema validation

### CODE REVIEW

File: pkg/config/transport_test.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+			t.Fatal(err)
+		}
+
+		if result.Valid() {
+			t.Errorf("have %v, want %v", input, want)
+		}
+	}
+}
+
+func TestDecodeJSONSchemaEmptyBody(t *testing.T) {
```
Comment (xla) [inline pkg/config/transport_test.go]: Great coverage on the possible failures. We should also have a test case that succeeds so we are confident the middleware works for the positive case.

File: pkg/config/schema.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package config
+
+import "github.com/xeipuuv/gojsonschema"
+
+var decodeClientPayloadSchema *gojsonschema.Schema
+
+const requestCapabilities = `
```
Comment (xla) [inline pkg/config/schema.go]: Usually consts are declared after vars in the rest of codebase.

File: pkg/config/transport_test.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "en_GB"}, "os": {"platform": "WatchOS", "version": "9.4"}}}`,                                          // TimezoneOffset missing
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "en_GB", "timezoneOffset": 7200}}}`,                                                                   // OS missing
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "en_GB", "timezoneOffset": 7200}, "os": {}}}`,                                                         // Empty Os object
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "en_GB", "timezoneOffset": 7200}, "os": {"platform": "WatchOS"}}}`,                                    // Version missing
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "testLocale", "timezoneOffset": 7201}, "os": {"platform": "WatchOS", "version": "9.4"}}}`,             // Wrong locale format
+	`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "testLocale", "timezoneOffset": 7201}, "os": {"platform": "WatchOS", "version": "9.4"}}, "user": {}}`, // Empty User object
+}
+
+func TestPayloadValidation(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (xla) [reply pkg/config/transport_test.go]: Maybe we can also add one case which succeeds against `requestCapabilities`.

File: pkg/config/schema.go (hunk 2 above)
Comment (alexsavu) [reply pkg/config/schema.go]: @xla I think you mean before. but you're right

File: pkg/config/transport_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexsavu) [reply pkg/config/transport_test.go]: although obvious and I should have picked up on it this is a very good point

File: pkg/config/schema.go (hunk 2 above)
Comment (xla) [reply pkg/config/schema.go]: Ooops, yes correctly.

File: pkg/config/schema_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+	testSuccessInputs = []string{
+		`{"app": {"version": "6.4.1"}, "device": {"location": {"locale": "en_GB", "timezoneOffset": 7201}, "os": {"platform": "WatchOS", "version": "9.4"}}, "user": {"age": 23}}`, // Working case
+	}
+)
+
+func TestPayloadValidationFailure(t *testing.T) {
+	var (
+		want = "invalid JSON error"
+	)
```
Comment (xla) [inline pkg/config/schema_test.go]: For a single variable it's almost always preferrable to use the short assignment (`:=`) as this takes up too much space for no gain.

File: pkg/config/schema_test.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		if result.Valid() {
+			t.Errorf("have %v, want %v", input, want)
+		}
+	}
+}
+
+func TestPayloadValidationSuccess(t *testing.T) {
+	var (
+		want = "valid JSON input"
+	)
```
Comment (xla) [inline pkg/config/schema_test.go]: Same as above.

File: pkg/config/schema_test.go (hunk 4 above)
Comment (alexsavu) [reply pkg/config/schema_test.go]: of course.  the compile error makes sense now. I was missing the `:=`

**Pull request merged**